refactor(contracts): Make FreezeVotingBase truly abstract for broader inheritance

adamgall opened the pull request:
Depends on https://github.com/decentdao/decent-contracts/pull/421

Fixes [ENG-1271](https://linear.app/decent-labs/issue/ENG-1271/refactor-freezevotingbase-to-support-standalone-freeze-voting)

## Motivation

The current FreezeVotingBase implementation contains ownership logic and an `unfreeze()` function that prevents FreezeVotingStandaloneV1 from inheriting from it. This forces standalone freeze voting to duplicate ~50 lines of core freeze voting logic, violating DRY principles and making maintenance harder. By making FreezeVotingBase truly abstract, we enable clean inheritance for both owned (parent/child) and self-governed (standalone) freeze voting implementations.

## Change Summary

- Remove `unfreeze()` function and Ownable2StepUpgradeable from FreezeVotingBase
- Move ownership logic to FreezeVotingMultisigV1 and FreezeVotingAzoriusV1
- Update FreezeVotingStandaloneV1 to inherit from FreezeVotingBase
- Remove duplicate freeze voting logic from standalone implementation
- Update interfaces to reflect the new inheritance structure

# Project Plan

## Current State Analysis

### Pain Points

1. **Code Duplication**: FreezeVotingStandaloneV1 duplicates core functionality:

   - `_initializeFreezeVote()` - Proposal initialization logic
   - `_recordFreezeVote()` - Vote recording and threshold checking
   - `isFrozen()` and `lastKnownFreezeTime()` - State getters
   - LightAccountValidator integration

2. **Inheritance Blockers**:

   - FreezeVotingBase contains `onlyOwner unfreeze()` function
   - Inherits from Ownable2StepUpgradeable
   - Assumes external ownership model incompatible with self-governance

3. **Maintenance Burden**:
   - Changes to freeze voting logic must be duplicated
   - Risk of implementations diverging over time
   - Harder to ensure consistent behavior

## Goals and Success Criteria

### Primary Goals

1. Enable FreezeVotingStandaloneV1 to inherit from FreezeVotingBase
2. Eliminate all code duplication between base and implementations
3. Maintain backward compatibility for existing contracts
4. Create clean, maintainable inheritance hierarchy

### Success Criteria

- [x] All duplicate code removed from FreezeVotingStandaloneV1
- [x] FreezeVotingBase is truly abstract (no ownership assumptions)
- [x] All existing tests pass without modification
- [x] Clear separation of concerns between base and implementations
- [x] No breaking changes to external interfaces

## Implementation Steps

1. **Refactor FreezeVotingBase**

   - Remove `unfreeze()` function from contract
   - Remove `unfreeze()` from IFreezeVotingBase interface
   - Remove Ownable2StepUpgradeable inheritance
   - Update `__FreezeVotingBase_init` to remove owner parameter
   - Update NatSpec documentation to reflect abstract nature

2. **Update FreezeVotingMultisigV1**

   - Add `import` for Ownable2StepUpgradeable
   - Add Ownable2StepUpgradeable to inheritance list
   - Add `__Ownable_init(owner_)` to initializer
   - Implement `unfreeze()` with `onlyOwner` modifier
   - Add `unfreeze()` to IFreezeVotingMultisigV1 interface
   - Mark `unfreeze()` as `override` in implementation

3. **Update FreezeVotingAzoriusV1**

   - Apply same changes as FreezeVotingMultisigV1
   - Ensure strategy clearing logic remains in `unfreeze()`
   - Add `unfreeze()` to IFreezeVotingAzoriusV1 interface

4. **Update FreezeVotingBase Tests**

   - Update mock implementation to add ownership
   - Fix initialization calls to match new signature
   - Ensure all base functionality tests still pass

5. **Refactor FreezeVotingStandaloneV1**

   - Change inheritance from `IFreezeVotingBase` to `FreezeVotingBase`
   - Remove duplicate implementations:
     - `_initializeFreezeVote()`
     - `_recordFreezeVote()` - `isFrozen()` - `lastKnownFreezeTime()`
   - Update constructor to call base `_disableInitializers()`
   - Call `__FreezeVotingBase_init` in initializer
   - Keep standalone-specific `unfreeze()` implementation

6. **Final Testing and Validation**
   - Run full test suite: `npm test`
   - Verify no compilation warnings
   - Check upgrade compatibility
   - Run linting and formatting

## Benefits

- **Reduced Maintenance**: Single source of truth for freeze voting logic
- **Better Testability**: Core logic tested once in base contract
- **Cleaner Architecture**: Clear separation between base functionality and access control
- **Future Flexibility**: Easy to add new freeze voting variants
- **Type Safety**: Proper inheritance ensures interface compatibility

## Progress Tracking

- [x] Refactor FreezeVotingBase to remove ownership
- [x] Update IFreezeVotingBase interface
- [x] Add ownership to FreezeVotingMultisigV1
- [x] Add ownership to FreezeVotingAzoriusV1
- [x] Update both interfaces with unfreeze function
- [x] Refactor FreezeVotingStandaloneV1 to inherit from base
- [x] Fix all compilation errors
- [x] Update and run all tests
- [x] Remove freezeVotedAt/unfreezeVotedAt mappings (user feedback)
- [x] Add NoVotes check to castUnfreezeVote
- [x] Remove UnfreezeProposal struct (user feedback)
- [x] DRY up \_getVotes and \_getUnfreezeVotes with shared function

## Notes and Decisions

- **Decision**: Made FreezeVotingBase abstract rather than creating new intermediate base
  - Rationale: Simpler hierarchy, less code churn, maintains conceptual clarity
- **Decision**: Keep LightAccountValidator in base rather than duplicating
  - Rationale: All freeze voting variants benefit from gasless voting
- **Discovery**: Test initialization required updating due to parameter changes

  - Resolution: Added lightAccountFactory parameter to test setup

- **User Feedback**: Removed excessive voting restrictions

  - Removed `freezeVotedAt` and `unfreezeVotedAt` mappings
  - Now relies solely on VoteTracker for double-vote prevention
  - Allows voters with multiple configs to use all of them

- **User Feedback**: Simplified storage structure

  - Removed `UnfreezeProposal` struct
  - Moved properties directly to main storage struct
  - Reduces nesting and improves readability

- **Refactoring**: Created shared `_aggregateVotes` function
  - Eliminates ~60 lines of duplicate code
  - Single place to maintain voting aggregation logic
  - Clearer separation between freeze and unfreeze voting

tbwebb22 commented on contracts/deployables/freeze-voting/FreezeVotingAzoriusV1.sol:
```diff
@@ -227,25 +231,20 @@ contract FreezeVotingAzoriusV1 is
+        $base.freezeProposalVoteCount = 0;
+
         // Clear the strategy snapshot to ensure fresh capture next time
         $.freezeProposalStrategy = address(0);
```
Do we need this line?

adamgall replied:
as we spoke about, we probably don't "need" it. but maybe it's "nice to have"?

if we removed it, that means that in the periods of time in which a freeze was _not_ in effect, this property would hold the strategy address that was used in the _last_ freeze, which is just stale.

i personally think it's worth the extra SSTORE (of the zero address, which is cheaper than a non-zero SSTORE) to clean up that variable.

tbwebb22 replied:
ok that makes sense, I'm cool with keeping the reset

adamgall committed:
refactor(contracts): Make FreezeVotingBase truly abstract for broader inheritance

## Motivation

The current FreezeVotingBase implementation contains ownership logic and an `unfreeze()` function that prevents FreezeVotingStandaloneV1 from inheriting from it. This forces standalone freeze voting to duplicate ~50 lines of core freeze voting logic, violating DRY principles and making maintenance harder. By making FreezeVotingBase truly abstract, we enable clean inheritance for both owned (parent/child) and self-governed (standalone) freeze voting implementations.

## Change Summary

- Remove `unfreeze()` function and Ownable2StepUpgradeable from FreezeVotingBase
- Move ownership logic to FreezeVotingMultisigV1 and FreezeVotingAzoriusV1
- Update FreezeVotingStandaloneV1 to inherit from FreezeVotingBase
- Remove duplicate freeze voting logic from standalone implementation
- Update interfaces to reflect the new inheritance structure

# Project Plan

## Current State Analysis

### Pain Points

1. **Code Duplication**: FreezeVotingStandaloneV1 duplicates core functionality:

   - `_initializeFreezeVote()` - Proposal initialization logic
   - `_recordFreezeVote()` - Vote recording and threshold checking
   - `isFrozen()` and `lastKnownFreezeTime()` - State getters
   - LightAccountValidator integration

2. **Inheritance Blockers**:

   - FreezeVotingBase contains `onlyOwner unfreeze()` function
   - Inherits from Ownable2StepUpgradeable
   - Assumes external ownership model incompatible with self-governance

3. **Maintenance Burden**:
   - Changes to freeze voting logic must be duplicated
   - Risk of implementations diverging over time
   - Harder to ensure consistent behavior

## Goals and Success Criteria

### Primary Goals

1. Enable FreezeVotingStandaloneV1 to inherit from FreezeVotingBase
2. Eliminate all code duplication between base and implementations
3. Maintain backward compatibility for existing contracts
4. Create clean, maintainable inheritance hierarchy

### Success Criteria

- [ ] All duplicate code removed from FreezeVotingStandaloneV1
- [ ] FreezeVotingBase is truly abstract (no ownership assumptions)
- [ ] All existing tests pass without modification
- [ ] Clear separation of concerns between base and implementations
- [ ] No breaking changes to external interfaces

## Implementation Steps

1. **Refactor FreezeVotingBase**

   - Remove `unfreeze()` function from contract
   - Remove `unfreeze()` from IFreezeVotingBase interface
   - Remove Ownable2StepUpgradeable inheritance
   - Update `__FreezeVotingBase_init` to remove owner parameter
   - Update NatSpec documentation to reflect abstract nature

2. **Update FreezeVotingMultisigV1**

   - Add `import` for Ownable2StepUpgradeable
   - Add Ownable2StepUpgradeable to inheritance list
   - Add `__Ownable_init(owner_)` to initializer
   - Implement `unfreeze()` with `onlyOwner` modifier
   - Add `unfreeze()` to IFreezeVotingMultisigV1 interface
   - Mark `unfreeze()` as `override` in implementation

3. **Update FreezeVotingAzoriusV1**

   - Apply same changes as FreezeVotingMultisigV1
   - Ensure strategy clearing logic remains in `unfreeze()`
   - Add `unfreeze()` to IFreezeVotingAzoriusV1 interface

4. **Update FreezeVotingBase Tests**

   - Update mock implementation to add ownership
   - Fix initialization calls to match new signature
   - Ensure all base functionality tests still pass

5. **Refactor FreezeVotingStandaloneV1**

   - Change inheritance from `IFreezeVotingBase` to `FreezeVotingBase`
   - Remove duplicate implementations:
     - `_initializeFreezeVote()`
     - `_recordFreezeVote()`
     - `isFrozen()`
     - `lastKnownFreezeTime()`
   - Update constructor to call base `_disableInitializers()`
   - Call `__FreezeVotingBase_init` in initializer
   - Keep standalone-specific `unfreeze()` implementation

6. **Final Testing and Validation**
   - Run full test suite: `npm test`
   - Verify no compilation warnings
   - Check upgrade compatibility
   - Run linting and formatting

## Benefits

- **Reduced Maintenance**: Single source of truth for freeze voting logic
- **Better Testability**: Core logic tested once in base contract
- **Cleaner Architecture**: Clear separation between base functionality and access control
- **Future Flexibility**: Easy to add new freeze voting variants
- **Type Safety**: Proper inheritance ensures interface compatibility

## Progress Tracking

- [x] Refactor FreezeVotingBase to remove ownership
- [x] Update IFreezeVotingBase interface
- [x] Add ownership to FreezeVotingMultisigV1
- [x] Add ownership to FreezeVotingAzoriusV1
- [x] Update both interfaces with unfreeze function
- [x] Refactor FreezeVotingStandaloneV1 to inherit from base
- [x] Fix all compilation errors
- [x] Update and run all tests
- [x] Remove freezeVotedAt/unfreezeVotedAt mappings (user feedback)
- [x] Add NoVotes check to castUnfreezeVote
- [x] Remove UnfreezeProposal struct (user feedback)
- [x] DRY up \_getVotes and \_getUnfreezeVotes with shared function

## Notes and Decisions

- **Decision**: Made FreezeVotingBase abstract rather than creating new intermediate base
  - Rationale: Simpler hierarchy, less code churn, maintains conceptual clarity
- **Decision**: Keep LightAccountValidator in base rather than duplicating
  - Rationale: All freeze voting variants benefit from gasless voting
- **Discovery**: Test initialization required updating due to parameter changes

  - Resolution: Added lightAccountFactory parameter to test setup

- **User Feedback**: Removed excessive voting restrictions

  - Removed `freezeVotedAt` and `unfreezeVotedAt` mappings
  - Now relies solely on VoteTracker for double-vote prevention
  - Allows voters with multiple configs to use all of them

- **User Feedback**: Simplified storage structure

  - Removed `UnfreezeProposal` struct
  - Moved properties directly to main storage struct
  - Reduces nesting and improves readability

- **Refactoring**: Created shared `_aggregateVotes` function
  - Eliminates ~60 lines of duplicate code
  - Single place to maintain voting aggregation logic
  - Clearer separation between freeze and unfreeze voting

adamgall committed:
Merge branch 'develop' into refactor-freezevotingbase
